src/web/transactions.rs

```diff
@@ -3,122 +3,103 @@ use std::collections::BTreeSet;
-                    payload,
-                    signatures,
-                    rejection_reason,
-                } = tx;
-                let base = TransactionBase::new(hash, payload, signatures)
-                    .wrap_err("Failed to make TransactionBase")?;
-                Ok(Self::Rejected(RejectedTransactionDTO {
-                    base,
-                    rejection_reason: rejection_reason.into(),
-                }))
-            }
-        }
+    fn try_from(tx_result: TransactionQueryResult) -> Result<Self> {
+        let TransactionValue { tx, error } = tx_result.transaction();
+        let block_hash = tx_result.block_hash();
+
+        Self::new(
+            tx.hash(),
+            block_hash,
+            tx.payload(),
+            tx.signatures().clone(),
+            error.clone(),
+        )
+        .wrap_err("Failed to make TransactionDTO")
     }
 }
 
-#[derive(Serialize)]
-struct TransactionBase {
-    hash: SerScaleHex<HashOf<SignedTransaction>>,
-    block_hash: SerScaleHex<Hash>,
-    payload: TransactionPayloadDTO,
-    signatures: BTreeSet<SignatureDTO>,
-}
-
-impl TransactionBase {
-    fn new<T, U>(hash: HashOf<SignedTransaction>, payload: Payload, signatures: T) -> Result<Self>
-    where
-        T: IntoIterator<Item = U>,
-        U: Into<Signature>,
-    {
+impl TransactionDTO {
+    fn new(
+        hash: HashOf<VersionedSignedTransaction>,
+        block_hash: &HashOf<CommittedBlock>,
+        payload: &TransactionPayload,
+        signatures: SignaturesOf<TransactionPayload>,
+        rejection_reason: Option<TransactionRejectionReason>,
+    ) -> Result<Self> {
         Ok(Self {
             hash: hash.into(),
-
-            // FIXME https://github.com/hyperledger/iroha/issues/2301
-            block_hash: Hash::zeroed().into(),
-
+            block_hash: (*block_hash).into(),
             payload: payload.try_into().wrap_err("Failed to map Payload")?,
             signatures: signatures
                 .into_iter()
                 .map(Into::<Signature>::into)
                 .map(Into::into)
                 .collect(),
+            rejection_reason: rejection_reason.map(SerScaleHex),
         })
     }
 }
 
-#[derive(Serialize)]
-pub struct CommittedTransactionDTO {
-    #[serde(flatten)]
-    base: TransactionBase,
-}
-
-/// Just as [`CommittedTransactionDTO`], but with rejection reason
-#[derive(Serialize)]
-pub struct RejectedTransactionDTO {
-    #[serde(flatten)]
-    base: TransactionBase,
-    rejection_reason: SerScaleHex<TransactionRejectionReason>,
-}
-
 #[derive(Serialize)]
 pub struct TransactionPayloadDTO {
     account_id: String,
     instructions: ExecutableDTO,
     creation_time: Timestamp,
-    time_to_live_ms: u32,
-    nonce: Option<u32>,
+    time_to_live_ms: Option<NonZeroU64>,
+    nonce: Option<NonZeroU32>,
     metadata: UnlimitedMetadata,
 }
 
-impl TryFrom<Payload> for TransactionPayloadDTO {
+impl TryFrom<&TransactionPayload> for TransactionPayloadDTO {
```

Comment:
I think it is not necessary to borrow at this stage of development. You can do simply this:

```suggestion
impl TryFrom<TransactionPayload> for TransactionPayloadDTO {
```

Reply:
Okay. Then, resolve this conversation when you push the related changes.